DB/src/main/java/io/deephaven/db/v2/by/StreamFirstOrLastByChunkedOperator.java

```diff
@@ -0,0 +1,142 @@
+     * and {@link #outputColumns}.
+     */
+    protected final int numResultColumns;
+    /**
+     * Result columns, parallel to {@link #inputColumns} and {@link #outputColumns}.
+     */
+    protected final Map<String, ArrayBackedColumnSource<?>> resultColumns;
+    /**
+     * <p>Input columns, parallel to {@link #outputColumns} and {@link #resultColumns}.
+     * <p>These are the source columns from the upstream table, reinterpreted to primitives where applicable.
+     */
+    protected final ColumnSource<?>[] inputColumns;
+    /**
+     * <p>Output columns, parallel to {@link #inputColumns} and {@link #resultColumns}.
+     * <p>These are the result columns, reinterpreted to primitives where applicable.
+     */
+    protected final WritableSource<?>[] outputColumns;
+    /**
+     * Cached pointer to the most-recently allocated {@link #redirections}.
+     */
+    protected SoftReference<LongArraySource> cachedRedirections;
+    /**
+     * Map from destination slot to first key. Only used during a step to keep track of the appropriate rows to copy into
+     * the output columns.
+     */
+    protected LongArraySource redirections;
+
+    protected StreamFirstOrLastByChunkedOperator(@NotNull final MatchPair[] resultPairs, @NotNull final Table streamTable) {
+        numResultColumns = resultPairs.length;
+        inputColumns = new ColumnSource[numResultColumns];
+        outputColumns = new WritableSource[numResultColumns];
+        final Map<String, ArrayBackedColumnSource<?>> resultColumnsMutable = new LinkedHashMap<>(numResultColumns);
+        for (int ci = 0; ci < numResultColumns; ++ci) {
+            final MatchPair resultPair = resultPairs[ci];
+            final ColumnSource<?> streamSource = streamTable.getColumnSource(resultPair.left());
+            final ArrayBackedColumnSource<?> resultSource = ArrayBackedColumnSource.getMemoryColumnSource(0, streamSource.getType(), streamSource.getComponentType());
+            resultColumnsMutable.put(resultPair.left(), resultSource);
+            inputColumns[ci] = ReinterpretUtilities.maybeConvertToPrimitive(streamSource);
+            // Note that ArrayBackedColumnSources implementations reinterpret very efficiently where applicable.
+            outputColumns[ci] = (WritableSource<?>) ReinterpretUtilities.maybeConvertToPrimitive(resultSource);
+            Assert.eq(inputColumns[ci].getChunkType(), "inputColumns[ci].getChunkType()", outputColumns[ci].getChunkType(), "outputColumns[ci].getChunkType()");
+        }
+        resultColumns = Collections.unmodifiableMap(resultColumnsMutable);
+        cachedRedirections = new SoftReference<>(redirections = new LongArraySource());
+    }
+
+    @Override
+    public final Map<String, ? extends ColumnSource<?>> getResultColumns() {
+        return resultColumns;
+    }
+
+    @Override
+    public final void startTrackingPrevValues() {
+        Arrays.stream(outputColumns).forEach(ColumnSource::startTrackingPrevValues);
```

Comment:
For firstBy I think this is unnecessary, because the columns will never change after you write a value down.

Reply:
To clarify for posterity: this makes sense because it's undefined to read previous values for rows that weren't present on the previous cycle.